feat: defi

jpsains opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Brings DeFi positions to extension.

Integrates Metamask's DeFi API which supports the following protocols Aave, Curve, Compound, Uniswap, Lido plus many more.

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/31751?quickstart=1)

## **Related issues**

Implements: https://consensyssoftware.atlassian.net/browse/MMASSETS-554

## **Manual testing steps**

1. Click on DeFi tab
2. Click on DeFi position
3. View DeFi Details screen

## **Screenshots/Recordings**

Recording: https://www.loom.com/share/b101241f877e4cbb81e2810417dc5fc9?sid=93379854-931d-408d-8b86-922606894151


<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

![Google Chrome 2025-04-09 15 38 10](https://github.com/user-attachments/assets/838492c9-f109-4263-bce4-daa51458616f)
![Google Chrome 2025-04-09 15 38 19](https://github.com/user-attachments/assets/a4b14153-0199-4c53-88ec-6169fdeac560)



## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

jpsains committed:
feat: defi

jpsains committed:
feat: defi tab

jpsains committed:
feat: lint fix

jpsains committed:
feat: unused imports

jpsains committed:
feat: improve comment

jpsains commented on .yarn/patches/@metamask-assets-controllers-npm-55.0.1-c79349f4b7.patch:
Patching in the unmerged defi controller - not to be merged

The plan is to merge defi controller remove then remove the defi controller patch

jpsains committed:
feat: fix

jpsains commented on .yarn/patches/@metamask-assets-controllers-npm-55.0.1-c79349f4b7.patch:
need to re-add patch back in when defi position controller is merged

jpsains commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies

jpsains committed:
feat: new defi api

jpsains commented on .yarn/patches/@metamask-previews-assets-controllers-npm-56.0.0-preview-9d7b1fa6-8481204834.patch:
This is the same patch as .yarn/patches/@metamask-assets-controllers-npm-55.0.1-c79349f4b7.patch

(it looks different because of prettier)

jpsains commented on package.json:
```diff
@@ -259,7 +259,7 @@
     "@metamask/announcement-controller": "^7.0.3",
     "@metamask/approval-controller": "^7.0.0",
-    "@metamask/assets-controllers": "patch:@metamask/assets-controllers@npm%3A55.0.1#~/.yarn/patches/@metamask-assets-controllers-npm-55.0.1-c79349f4b7.patch",
+    "@metamask/assets-controllers": "npm:@metamask-previews/assets-controllers@56.0.0-preview-9d7b1fa6",
```
We need to merge defi controller first

jpsains commented on ui/components/app/assets/token-cell/token-cell.tsx:
```diff
@@ -74,39 +72,9 @@ export default function TokenCell({
-    },
-    [onClick, token, showScamWarningModal, trackEvent],
-  );
-
```
this code has been moved to token-cell

jpsains commented on ui/components/multichain/account-overview/account-overview-eth.tsx:
```diff
@@ -1,15 +1,19 @@
 export type AccountOverviewEthProps = AccountOverviewCommonProps;
 
 export const AccountOverviewEth = (props: AccountOverviewEthProps) => {
+  const defiPositionsEnabled = useSelector(getIsDefiPositionsEnabled);
   return (
     <AccountOverviewLayout
       showTokens={true}
       showNfts={true}
+      showDefi={defiPositionsEnabled}
       showActivity={true}
       {...props}
```
Feature flag

jpsains replied:
Feature flag

jpsains commented on ui/pages/defi/components/defi-details-list.tsx:
```diff
@@ -0,0 +1,150 @@
+  (typeof PositionTypeLabels)[keyof typeof PositionTypeLabels];
+export type PositionTypeKeys = keyof typeof PositionTypeLabels;
+
+type Protocols = GroupedPositions['protocols'];
+type Protocol = Protocols[keyof GroupedPositions['protocols']];
+type ProtocolPositionTypes = Protocol['positionTypes'];
+type PositionGroup = NonNullable<
+  ProtocolPositionTypes[keyof ProtocolPositionTypes]
+>;
+export type ProtocolTokenWithMarketValue = NonNullable<
+  PositionGroup['positions'][0][0]
+>;
+type UnderlyingWithMarketValue = NonNullable<
+  ProtocolTokenWithMarketValue['tokens']
+>[0];
```
Nice to have these types exported from defi controller

jpsains commented on lavamoat/browserify/flask/policy.json:
```diff
@@ -1068,10 +1068,12 @@
         "Headers": true,
         "URL": true,
         "URLSearchParams": true,
+        "__import__": true,
```
@bergarces fyi we have some policy changes on @metamask/assets-controllers

The fetch Im expecting because we need to fetch to GET defi-positions from defi-api

`"__import__": true,` I dont see this permission on other packages - know much about it?

cc. @darkwing

shane-t replied:
I think it's from this dynamic import https://github.com/MetaMask/core/blob/deaf2dc70d8f27a8470f759be8bc911f29d40016/packages/assets-controllers/src/assetsUtil.ts#L260

bergarces replied:
If we restore the patch that replaces the dynamic import this won't be added.

jpsains committed:
feat: remove log

jpsains commented on ui/components/app/assets/token-cell/token-cell.tsx:
```diff
@@ -74,39 +72,9 @@ export default function TokenCell({
 
   const tokenDisplayInfo = useTokenDisplayInfo({
     token,
+    fixCurrencyToUSD,
```
when defi supports local currency settings we can remove this

bergarces committed:
use new controller preview

bergarces committed:
patching

bergarces committed:
lavamoat

bergarces committed:
patch current assets-controllers version

bergarces committed:
remove patch

bergarces commented on lavamoat/browserify/main/policy.json:
```diff
@@ -1072,6 +1072,7 @@
         "clearTimeout": true,
         "console.error": true,
         "console.log": true,
+        "fetch": true,
```
Expected, since we are fetching from an API in the new DeFiPositionsController.

bergarces committed:
Merge branch 'main' into mm-defi-positions

bergarces committed:
correct type name

bergarces committed:
circular deps

bergarces commented:
> You have added a circular dependency here, which is not okay.

It's a Draft PR, it'll get sorted before it's ready to be merged.

bergarces committed:
removed circular dep

bergarces committed:
added defi api to list of hosts

bergarces committed:
snapshot updates

bergarces committed:
correct snapshots

bergarces committed:
Merge branch 'main' into mm-defi-positions

bergarces committed:
Merge branch 'main' into mm-defi-positions

darkwing commented on privacy-snapshot.json:
```diff
@@ -86,5 +86,8 @@
-  "www.4byte.directory"
+  "www.4byte.directory",
+  "api.web3modal.org",
+  "www.4byte.directory",
```
Looks like 4byte is added twice (?)

bergarces replied:
Somehow made a mess when merging. I've cleared it out.

bergarces committed:
remove entries

bergarces committed:
Merge branch 'main' into mm-defi-positions

bergarces committed:
Merge branch 'main' into mm-defi-positions

bergarces committed:
restore value

jpsains committed:
feat: fix gb locales lint error

pnarayanaswamy committed:
test: add integration tests for defi feature (#31998)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/31998?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

Co-authored-by: jpsains <32621022+jpsains@users.noreply.github.com>

bergarces committed:
repatch with new changes

bergarces committed:
Merge branch 'main' into mm-defi-positions

jpsains committed:
feat: e2e tests

jpsains committed:
Merge branch 'main' into mm-defi-positions

jpsains committed:
feat: drop console logs

jpsains committed:
feat: confict

jpsains commented:
@SocketSecurity ignore npm/@metamask/assets-controllers@57.0.0

bergarces committed:
feature flag

jpsains committed:
feat: defi

jpsains committed:
feat: fix lint

jpsains committed:
feat: unit tests

jpsains committed:
feat: merge

jpsains committed:
feat: update snapshots

jpsains committed:
feat: lint

jpsains committed:
feat: lint

jpsains committed:
Merge branch 'main' into mm-defi-positions

gambinish commented on ui/components/app/assets/asset-list/asset-list-control-bar/asset-list-control-bar.tsx:
```diff
@@ -233,10 +235,12 @@ const AssetListControlBar = ({
-            showTokensLinks={showTokensLinks}
-            onClick={toggleImportTokensPopover}
-          />
+          {showImportControlls && (
```
Small typo:

```suggestion
        {showImportControls && (
```

gambinish commented on ui/components/app/assets/defi-list/defi-list.tsx:
```diff
@@ -0,0 +1,259 @@
+    return [];
+  }
+
+  return Object.values(protocolPositions.positionTypes).flatMap(
+    (displayTokens) =>
+      displayTokens?.positions?.flatMap(
+        (nestedToken) =>
+          nestedToken?.flatMap(
+            (token) =>
+              token?.tokens?.map((underlying) => ({
+                symbol: underlying?.symbol || '',
+                avatarValue: underlying?.iconUrl || '',
+              })) || [],
+          ) || [],
+      ) || [],
+  );
+};
```
Can this somehow be put into a memoized selector somehow? I'm worried that these nested loops may cause unnecessary re-renders.

If not a selector, maybe we can put this into a `useMemo` hook to only get evaluated when we expect.

jpsains replied:
was being memo'd but made more clear

gambinish commented on ui/components/app/assets/defi-list/defi-list.tsx:
```diff
@@ -0,0 +1,259 @@
+      ) : (
+        <TokenCell
+          key={`empty-defi-list`}
+          token={{
+            address: '0x',
+            title: 'Start earning',
+            symbol: 'Start earning',
+            tokenFiatAmount: 0,
+            image: `images/logo/metamask-fox.svg`,
+            primary: '0',
+            secondary: 0,
+            decimals: 10,
+            chainId: '0x1',
+            isStakeable: true,
+          }}
```
Maybe we can save this object into a variable for clarity on what it represents.

Maybe `defiListFallbackToken` or something along those lines.

gambinish commented on ui/components/app/assets/token-cell/token-cell.tsx:
```diff
@@ -173,10 +149,12 @@ export default function TokenCell({
-              token={{ ...token, ...tokenDisplayInfo }}
-              privacyMode={privacyMode}
-            />
+            {primaryDisplayOverride?.() ?? (
+              <TokenCellPrimaryDisplay
+                token={{ ...token, ...tokenDisplayInfo }}
+                privacyMode={privacyMode}
+              />
+            )}
```
Could you explain what this condition is doing?

gambinish commented on ui/components/app/assets/token-list/token-list.tsx:
```diff
@@ -78,14 +84,34 @@ function TokenList({ onTokenClick }: TokenListProps) {
           token={token}
           privacyMode={privacyMode}
-          onClick={onTokenClick}
+          onClick={handleTokenClick(token)}
```
```suggestion
          onClick={() => handleTokenClick(token)}
```

jpsains replied:
handleTokenClick(token) returns type ```()=> void``` already above proposal returns ```()=>()=>void```

gambinish commented on ui/pages/defi/components/defi-details-page.tsx:
```diff
@@ -0,0 +1,139 @@
+          isHidden={privacyMode}
+          length={SensitiveTextLength.Medium}
+        >
+          {'$'}
```
Is this meant to be hardcoded?

jpsains replied:
Yes for now we only support dollar

gambinish commented on ui/pages/defi/components/defi-details-page.tsx:
```diff
@@ -0,0 +1,139 @@
+          length={SensitiveTextLength.Medium}
+        >
+          {'$'}
+          {protocolPosition.aggregatedMarketValue.toFixed(2)}
```
We could probably use `formatWithThreshold` here to format consistently. If we are limited to usd support for now, we can pass the currency as a hard coded value

gambinish commented on ui/pages/defi/components/defi-details-page.tsx:
```diff
@@ -0,0 +1,139 @@
+        <hr style={{ border: '1px solid var(--border-muted, #858B9A33)' }} />
+      </Box>
+      <Box display={Display.Flex} flexDirection={FlexDirection.Column}>
+        {['supply', 'borrow', 'stake', 'reward'].map((positionType) =>
```
Can we assign `['supply', 'borrow', 'stake', 'reward']` to a variable?

jpsains committed:
feat: fix

jpsains committed:
feat: feedback

jpsains committed:
feat: fix

jpsains committed:
feat: lint

jpsains committed:
feat: snapshot update

jpsains committed:
Merge branch 'main' into mm-defi-positions

jpsains committed:
feat: fix integration test

jpsains committed:
feat: conflicts

jpsains commented on privacy-snapshot.json:
```diff
@@ -27,6 +27,7 @@
   "configuration.metamask-institutional.io",
   "connect.trezor.io",
   "customnetwork.test",
+  "defiadapters.dev-api.cx.metamask.io",
```
URL checked by @jpsains and @bergarces and is expected

bergarces reviewed (approved):
Good job. We have a follow up ticket to refactor reused components, but this is a great fully functional and well scoped start point.
